content/docs/specs/native_histograms.md

```diff
@@ -0,0 +1,2422 @@
+use cases. (TODO: NHCB aren't fully merged into main as of now (2024-11-03).
+They are worked into this document as far as possible already. This information
+might not yet be relevant for released Prometheus versions.)
+
+### Buckets
+
+For standard schemas, buckets are represented as two lists, one for positive
+buckets and one for negative buckets. For custom buckets (schema -53), only the
+positive bucket list is used, but repurposed for all buckets.
+
+Any unpopulated buckets MAY be excluded from the lists. (Which is the reason
+why the buckets are often called _sparse buckets_.)
+
+For float histograms, the elements of the lists are float64 and
+represent the bucket population directly.
+
+For integer histograms, the elements of the lists are signed 64-bit integers
+(short: int64), and each element represents the bucket population as a delta to
+the previous bucket in the list. The first bucket in each list contains an
+absolute population (which can also be seen as a delta relative to zero).
+
+To map buckets in the lists to the indices as defined in the previous section,
+there are two lists of so-called _spans_, one for the positive buckets and one
+for the negative buckets.
+
+Each span consists of a pair of numbers, a signed 32-bit integer (short: int32)
+called _offset_ and an unsigned 32-bit integer (short: uint32) called _length_.
+Only the first span in each list can have a negative offset. It defines the
+index of the first bucket in its corresponding bucket list. (Note that for
+NHCBs, the index is always positive, see the [custom values
+section](#custom-values) below for details.) The length defines the number of
+consecutive buckets the bucket list starts with. The offsets of the following
+spans define the number of excluded (and thus unpopulated buckets). The lengths
+define the number of consecutive buckets in the list following the excluded
+buckets.
+
+The sum of all length values in each span list MUST be equal to the length of
+the corresponding bucket list.
+
+Empty spans (with a length of zero) are valid and MAY be used, although they
+are generally not useful and they SHOULD be eliminated by adding their offset
+to the offset of the following span. Similarly, spans that are not the first
+span in a list MAY have an offset of zero, although those offsets SHOULD be
+eliminated by adding their length to the previous span. Both cases are allowed
+so that producers of native histograms MAY pick whatever representation has the
+best resource trade-offs at that moment. For example, if a histogram is
+processed through various stages, it might be most efficient to only eliminate
+redundant spans after the last processing stage.
+
+In a similar spirit, there are situation where excluding every unpopulated
+bucket from the bucket list is most efficient, but in other situations, it
+might be better to reduce the number of spans by representing small numbers of
+unpopulated buckets explicitly.
```

Comment:
I don't fully understand here. How could we reduce the number of spans by representing small numbers of unpopulated buckets explicitly? Do you mean combine two successive spans into a new span with some unpopulated buckets in it?